go-apps/meep-wais/server/wais.go

```diff
@@ -357,7 +375,86 @@ func updateApInfo(name string, apMacId string, longitude *float32, latitude *flo
 		apId.Bssid = apMacId
 		apInfoComplete.ApId = apId
 		_ = rc.JSONSetEntry(baseKey+"AP:"+name, ".", convertApInfoCompleteToJson(&apInfoComplete))
-		checkAssocStaNotificationRegisteredSubscriptions(staMacIds, apMacId, false)
+		checkAssocStaNotificationRegisteredSubscriptions(staMacIds, apMacId, false, true)
+	}
+}
+
+func checkAssocStaPeriodTrigger() {
+
+	//loop through every subscriptions, lower period by one and invoke the notification if a check is warranted
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	if len(assocStaSubscriptionInfoMap) < 1 {
+		return
+	}
+
+	//decrease all subscriptions
+	//check all that applies
+	for _, subInfo := range assocStaSubscriptionInfoMap {
+		if subInfo != nil {
+			subInfo.NextTts--
+			//if no periodic check is needed, value is negative
+			if subInfo.NextTts != 0 {
+				continue
+			} else { //restart the nextTts and continue processing to send notification or not
+				subInfo.NextTts = subInfo.Subscription.NotificationPeriod
+			}
+		}
+	}
+	//find every AP info and reuse a function to store them all
+	var apInfoCompleteResp ApInfoCompleteResp
+	apInfoCompleteResp.ApInfoCompleteList = make([]ApInfoComplete, 0)
+	keyName := baseKey + "AP:*"
+	err := rc.ForEachJSONEntry(keyName, populateApInfoCompleteList, &apInfoCompleteResp)
+	if err != nil {
+		log.Error(err.Error())
+		return
+	}
+
+	//loop through the response for each AP and check subscription with no need for mutex (already used)
+	for _, apInfoComplete := range apInfoCompleteResp.ApInfoCompleteList {
+		checkAssocStaNotificationRegisteredSubscriptions(apInfoComplete.StaMacIds, apInfoComplete.ApId.Bssid, true, false)
+	}
+}
+
+func checkStaDataRatePeriodTrigger() {
+
+	//loop through every subscriptions, lower period by one and invoke the notification if a check is warranted
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	if len(staDataRateSubscriptionInfoMap) < 1 {
+		return
+	}
+
+	//decrease all subscriptions
+	//check all that applies
+	for _, subInfo := range staDataRateSubscriptionInfoMap {
+		if subInfo != nil {
+			subInfo.NextTts--
+			//if no periodic check is needed, value is negative
+			if subInfo.NextTts != 0 {
+				continue
+			} else { //restart the nextTts and continue processing to send notification or not
+				subInfo.NextTts = subInfo.Subscription.NotificationPeriod
```

Comment:
The code seems to do the right job for triggering periodic notifications, however it is difficult to follow/debug in some places. For example, subInfo.NextTts will cause a notification to be sent when it is set to max value. It may be clearer to add another flag in subInfo to indicate that period has expired in this case and that a notification may be sent if necessary.

Reply:
added a flag to make it clearer... but essentially same logic